### PR TITLE
[V1 API] paged property values (out) API

### PR DESCRIPTION
**Pull request by shifucun**
- Refactor the processor struct into "state" struct which allows composition for "in" and "out" cases.
- Consolidate to use one handler for both "in" and "out" cases.
- More abstraction of "state" functions for re-use.
- Change naming patter from "in_property_values" to "property_values_in" to match API URI

### CODE REVIEW

File: internal/server/v1/propertyvalues/common.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+func buildEmptyCursorGroup(n int) *pb.CursorGroup {
+	result := &pb.CursorGroup{Cursors: []*pb.Cursor{}}
+	for i := 0; i < n; i++ {
+		result.Cursors = append(result.Cursors, &pb.Cursor{ImportGroup: int32(i)})
+	}
+	return result
+}
+
+var action = func(jsonRaw []byte) (interface{}, error) {
```
Comment (pradh) [inline internal/server/v1/propertyvalues/common.go:35]: nit: action seems too general a name -  perhaps unmarshalAction?

File: proto/v1/property_values.proto [hunk 2]
```diff
@@ -37,11 +37,13 @@ message InPropertyValuesRequest {
   // [Optional]
   // The pagination token for getting the next set of entries. This is empty
   // for the first request and needs to be set in the subsequent request.
-  // This is the value returned from a prior call to InPropertyValuesResponse
+  // This is the value returned from a prior call to PropertyValuesResponse
   string next_token = 4;
+  // Direction can only be "in" and "out"
+  string direction = 5;
```
Comment (pradh) [inline proto/v1/property_values.proto:43]: Should we use enum (preferable), or boolean?

File: internal/server/v1/propertyvalues/property_values.go [hunk 3]
```diff
@@ -97,132 +78,123 @@ func InPropertyValues(
...
+			if !hasNext {
+				break
+			}
 		}
-	}
-	respToken := ""
-	for _, d := range proc.rawEntities {
-		// If any import group has more data, then should compute the token
-		if d != nil {
+		// If used import group has more data, then should compute the token
```
Comment (pradh) [inline internal/server/v1/propertyvalues/property_values.go:98]: Add some clarifying comment why this is only one import group, unlike "in"?

File: internal/server/v1/propertyvalues/state.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	state
+	// Min heap for entity merge sort
+	heap *entityHeap
+}
+
+type outState struct {
+	state
+	// The used import group (only one import group is used for out property values).
+	usedImportGroup int
+	//
```
Comment (pradh) [inline internal/server/v1/propertyvalues/state.go:51]: nit: delete line

File: internal/server/v1/propertyvalues/state.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	"container/heap"
+	"context"
+	"log"
+
+	cbt "cloud.google.com/go/bigtable"
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	"github.com/datacommonsorg/mixer/internal/store/bigtable"
+)
+
+type state struct {
```
Comment (pradh) [inline internal/server/v1/propertyvalues/state.go:27]: struct level comment?

File: internal/server/v1/propertyvalues/state.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+type inState struct {
+	state
+	// Min heap for entity merge sort
+	heap *entityHeap
+}
+
+type outState struct {
+	state
+	// The used import group (only one import group is used for out property values).
```
Comment (pradh) [inline internal/server/v1/propertyvalues/state.go:49]: I see, so this is how it is today?

File: internal/server/v1/propertyvalues/state.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+	// Store the raw cache data.
+	for idx, btData := range btDataList {
+		if data, ok := btData[entity]; ok {
+			pe := data.(*pb.PagedEntities)
+			s.rawEntities = append(s.rawEntities, pe.Entities)
```
Comment (pradh) [inline internal/server/v1/propertyvalues/state.go:94]: Should we append triples across multiple import-groups for out too? (assuming each distinct idx is a separate import-group)

File: internal/server/v1/propertyvalues/state.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+		ctx,
+		btGroup,
+		rowListMap,
+		action,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+	if data, ok := btDataList[importGroup][s.entity]; ok {
```
Comment (pradh) [inline internal/server/v1/propertyvalues/state.go:125]: assert !arcOut || importGroup == s.usedImportGroup ?

File: internal/server/v1/propertyvalues/state.go (hunk 5 above)
Comment (shifucun) [reply internal/server/v1/propertyvalues/state.go:29]: Done

File: internal/server/v1/propertyvalues/state.go (hunk 6 above)
Comment (shifucun) [reply internal/server/v1/propertyvalues/state.go:51]: Yep

File: internal/server/v1/propertyvalues/state.go (hunk 7 above)
Comment (shifucun) [reply internal/server/v1/propertyvalues/state.go:95]: This is the base case where only raw entities are saved. The rawEntities looks like
[
  [ig1_1, ig1_2, ..],
  [],
  [ig3_1, ig3_2]
]
There are no merging here.

File: internal/server/v1/propertyvalues/state.go (hunk 8 above)
Comment (shifucun) [reply internal/server/v1/propertyvalues/state.go:126]: Here "s" is the base struct that does not have usedImportGroup field. The caller func nextOut() in property_values.go could guarantee this condition.

File: proto/v1/property_values.proto (hunk 2 above)
Comment (shifucun) [reply proto/v1/property_values.proto:43]: This field is parsed by the ESP from the REST url where we could only get a string.

**Review by pradh — APPROVED** [review]

**Pull request merged**